src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -278,6 +284,17 @@ - (void)_handleSubscriptionEstablished
     // reset subscription attempt wait time when subscription succeeds
     _lastSubscriptionAttemptWait = 0;
 
+    // Check if estimated system start time can be made better
+    if (_estimatedStartTimeFromGeneralDiagnosticsUpTime) {
+        if (!_estimatedStartTime
+            || ([_estimatedStartTimeFromGeneralDiagnosticsUpTime compare:_estimatedStartTime] == NSOrderedAscending)) {
+            MTR_LOG_INFO("%@ SubscriptionEstablished: update estimated start time %@ => %@", self, _estimatedStartTime,
+                _estimatedStartTimeFromGeneralDiagnosticsUpTime);
+            _estimatedStartTime = _estimatedStartTimeFromGeneralDiagnosticsUpTime;
+        }
+        _estimatedStartTimeFromGeneralDiagnosticsUpTime = nil;
```

Comment:
So if our subscription drops we will clear the `_estimatedStartTime`, right?  But the resubscribe might well not get an updated `_estimatedStartTimeFromGeneralDiagnosticsUpTime` value, because we'll send a DataVersion for General Diagnostics and it might get optimized out.  Are we setting this to null to handle a case where a device switches from supporting the attribute to not supporting it across an OTA or something along those lines?

Reply:
The reason is really that we can't trust an older value to be valid on next subscription.

Reply:
Hmm.  So there are two cases here:

1. We resubscribe and get an updated value of UpTime.  In that case we should use that updated value.
2. We resubscribe and do not get an updated value of UpTime, but we got one before.  That almost certainly indicates it got optimized out and we should keep using the old one....

Reply:
> We resubscribe and do not get an updated value of UpTime, but we got one before. That almost certainly indicates it got optimized out and we should keep using the old one....

As an example, if OTA happened and UpTime is somehow no longer implemented, we can't trust a previous estimate, given the client has no visibility into whether UpTime-derived estimate is still valid after receiving a StartUp event during priming... and so we really have no choice but to abandon previous estimates at that time, unless we spec and implement some stronger requirement of UpTime to be reported after boot if it had previously been reported.

Reply:
Marking conversation resolved as the latest commit significantly changed logic.